### ISSUE TITLE
Syncode for fim models

### ISSUE THREAD

**Issue by AzizCode92** [uiuc-focal-lab/syncode#82]
First of all, thank you for this great work!
I would like to know if Syncode can also work with fill-in-the-middle models?
If yes, how?

**Comment by shubhamugare** [comment]
Hi @AzizCode92,

SynCode doesn't support FIM models at this point. Do you have any specific application in your mind?

**Comment by AzizCode92** [comment]
Thanks @shubhamugare.
Not really. I was just wondering how the parser would find the most probable next token if we have both prefix and suffix parts. 
Right now, it looks like the model is following the grammar by heart. It starts always by the first terminal I define in my grammar. 
IMH, I think it would take the prefix into consideration and produce the next token from there and at the same time respect the grammar rules.
It is a challenging task I guess 😅

**Comment by shubhamugare** [comment]
Yeah, it is definitely not straightforward extension of what currently exists within SynCode. In theory, one can use the constraints-based on suffix (in addition to the constraint from the prefix which SynCode currently uses) to find what is an appropriate stopping criterion for the MIM model.
